### PR TITLE
JAX: Random Forest – Train NAO brainwave model (CPU-only) (#320)

### LINKED ISSUE #320: JAX: Random Forest - train a brainwave model
Using the provided data, obtain the success rates for NAO control model with the brain. Use JAX Random Forest Classifier.

Check on Wiki - adjust to JAX accordingly:

Image

Do a pull request, adding the new JAX code implementation and trained model in the directory:
Image

**VERY IMPORTANT**
It is possible to implement a Random Forest algorithm in JAX without relying on scikit-learn. This approach involves building the decision trees and their ensemble logic directly using JAX's numerical computation capabilities.
Here are the key aspects of implementing a JAX-based Random Forest without scikit-learn:
Decision Tree Implementation: Each individual decision tree within the forest needs to be implemented from scratch. This involves defining functions for:
Splitting Criteria: Determining the best feature and split point at each node (e.g., Gini impurity for classification, mean squared error for regression).
Tree Construction: Recursively building the tree by splitting nodes until a stopping condition is met (e.g., maximum depth, minimum samples per leaf).
Prediction: Traversing the tree to make predictions for new data points.
Ensemble Logic: The Random Forest ensemble combines multiple decision trees. This requires:
Bootstrapping: Randomly sampling the training data with replacement to create multiple subsets for training individual trees.
Feature Randomness: Randomly selecting a subset of features at each split point within a tree.
Aggregation: Combining the predictions of individual trees (e.g., majority voting for classification, averaging for regression).
JAX Advantages:
Automatic Differentiation: JAX's grad function can be used if you need to optimize parameters within the tree construction process, though traditional Random Forests are not typically differentiable end-to-end.
JIT Compilation: Using jax.jit can significantly accelerate the execution of your custom Random Forest functions.
Vectorization (vmap): jax.vmap can be used to efficiently apply tree construction and prediction logic across multiple trees or data samples.
Device Acceleration: JAX allows you to run your implementation on GPUs or TPUs for faster computation.
Considerations:
Complexity: Implementing Random Forest from scratch is more complex than using scikit-learn's pre-built functions.
Optimization: Achieving optimal performance might require careful consideration of JAX's features and potential optimizations for your specific implementation.
Debugging: Debugging a custom implementation can be more challenging compared to using a well-tested library.
While it is a more involved process, implementing Random Forest in JAX without scikit-learn provides full control over the algorithm and allows for leveraging JAX's performance and flexibility.

_Reference:_
**Use Jax to create a Random Forest Classifier**
https://tracyrenee61.medium.com/use-jax-to-create-a-random-forest-classifier-0eccb3cc336f

### PR DESCRIPTION
**Pull request by Esma-Ali**
### Summary
Implements a CPU-safe JAX Random Forest Classifier for the NAO control brainwave model.

### Details
- Added `jax_rf.py`: CPU-friendly Random Forest using NumPy masking.
- Added `train_jax_rf.py` and `train_rf_cpu.py` for training.
- Added `combine_brainwaves.py` to merge OpenBCI EEG recordings into `brainwaves.csv`.
- Artifacts (metrics.json, confusion_matrix.png) stored in `deep-learning-prediction/JAX/artifacts/`.

### Results
- **Accuracy:** 99.96%
- **Macro F1:** 0.9996
- CPU-only mode enforced with:
  ```bash
  set JAX_PLATFORM_NAME=cpu
  set JAX_DISABLE_JIT=1

Addresses #320 
Closes #320

### CODE REVIEW

**Review by 3C-SCSU — APPROVED** [review]
Merging this newer version. Addresses: https://github.com/3C-SCSU/Avatar/issues/320

The older pull request is not merged.

**Comment by 3C-SCSU** [comment]
Jax: Random Forest implementation - do not use ScikiLearn. Use pure Jax.
